benchrun/python/benchrun/_benchmark.py

```diff
@@ -0,0 +1,280 @@
+import abc
+import copy
+import gc
+import multiprocessing as mp
+import time
+import uuid
+import warnings
+from typing import Any, Dict, List
+
+from benchadapt import BenchmarkResult
+
+from .cache import CacheManager
+from .case import CaseList
+
+
+class Iteration(abc.ABC):
+    """
+    Abstract class defining how to run one iteration of a benchmark.
+
+    Attributes
+    ----------
+
+    name : str
+        A name for the benchmark. Should be specified when inheriting.
+    cache : CacheManager
+        An CacheManager instance for clearing the disk cache when specified.
+        Do not mess with this.
+    env : dict
+        A dict stages can use to pass data between them.
+    """
+
+    name: str
+    cache = None
+    env: dict = None
+
+    def __init__(self) -> None:
+        assert self.name
+        self.cache = CacheManager()
+        self.env = {}
+
+    def setup(self, case: dict) -> None:
+        """
+        Global setup that runs once before any iteration.
+        """
+        pass
+
+    def before_each(self, case: dict) -> None:
+        """
+        Code to run in each iteration before timing.
+
+        Parameters
+        ----------
+        case : dict
+            A dict where keys are parameters and values are scalar arguments for a benchmark
+        """
+        pass
+
+    @abc.abstractmethod
+    def run(self, case: dict) -> None:
+        """
+        Code to time.
+
+        Parameters
+        ----------
+        case : dict
+            A dict where keys are parameters and values are scalar arguments for a benchmark
+        """
+
+    def after_each(self, case: dict) -> None:
+        """
+        Code to run in each iteration after timing.
+
+        Parameters
+        ----------
+        case : dict
+            A dict where keys are parameters and values are scalar arguments for a benchmark
+        """
+        pass
+
+    def teardown(self, case: dict) -> None:
+        """Global teardown that runs once after all iterations"""
+        pass
+
+    def __call__(
+        self, case: dict, iterations: int, settings: dict, queue: mp.Queue
+    ) -> dict:
+        """
+        Run iterations and return list of times.
+
+        Parameters
+        ----------
+        case : dict
+            A dict where keys are parameters and values are scalar arguments for a benchmark
+        iterations : int
+            Integer of repetitions to run
+        settings : dict
```

Comment:
Any reason these aren't explicit parameters? It would be nice to see all the settings explicitly

Reply:
They are explicit on `Benchmark`, which is where they should really be set, but to run this, they have to be passed through from `Benchmark` to `Iteration`, and it's just simpler to pass a dict than everything because with `multiprocessing` I've got to pass all the args through in a tuple:

```python
        proc = mp.Process(
            target=iteration, args=(case, iterations, self.settings, queue)
        )
```
which I kinda hate, and separating them out would exacerbate it. Maybe I could use `NamedTuple` or something? But the tl;dr is that users mostly shouldn't be calling this (though they can); they should use `Benchmark` as an interface instead.

Reply:
Yeah, that makes sense. I just read this first before I got to the more user-friendly code below.